### ISSUE TITLE
Update muon iso, trigger, ID SF for 2016 rereco.

### ISSUE THREAD

**Issue by bortigno** [UFLX2MuMu/Ntupliser#75]
Starting from #74, which does not work because I need to synchronise the plugins using the json and the cff that is passing root files. 

At the same time I realised I need to update the muon SF for rereco, so I will do that and fix the problem of #74 at the same time here.

**Comment by bortigno** [comment]
Working on branch `issue-75-muon-iso-id-trig-sf-2016-rereco`
Sister issue #55 and PR #58

**Comment by bortigno** [comment]
https://twiki.cern.ch/twiki/bin/view/CMS/MuonReferenceEffs2016LegacyRereco

https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies/blob/master/EfficienciesStudies/2016_legacy_rereco/jsonfiles/RunBCDEF_SF_ISO.json

https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies/blob/master/EfficienciesStudies/2016_legacy_rereco/jsonfiles/RunBCDEF_SF_ID.json

https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies/blob/master/EfficienciesStudies/2016_legacy_rereco/jsonfiles/RunGH_SF_ID.json

https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies/blob/master/EfficienciesStudies/2016_legacy_rereco/jsonfiles/RunGH_SF_ISO.json

**Comment by bortigno** [comment]
In #55 (2018) I decided to download the muon efficiency gitlab package and link to the Ntupliser. I think we can do the same here 

-  [] Need to update the README to include the download of the muon-POG efficiency gitlab package!

```
git clone https://gitlab.cern.ch/cms-muonPOG/MuonReferenceEfficiencies.git
```

then I need to change the cff pointinig to these new files

```
  Trig_eff_3_file  = cms.string("MuonReferenceEfficiencies/EfficienciesStudies/2016_trigger/EfficienciesAndSF_RunBtoF.root"),
MuID_eff_3_file  = cms.string("MuonReferenceEfficiencies/EfficienciesStudies/2016_legacy_rereco/jsonfiles/RunBCDEF_SF_ID.json"),
MuIso_eff_3_file = cms.string("MuonReferenceEfficiencies/EfficienciesStudies/2016_legacy_rereco/jsonfiles/RunBCDEF_SF_ISO.json")
```

but I also need the "GH" part so need to add the "_4" version of these.

I also added the "EMTF_bug" part.

For the moment I am adding the additional SF on a separate leaf of the tree and the combination will be done at a later stage. 

These are the files that I had to modify.
```
modified:   plugins/UFDiMuonsAnalyzer.cc
modified:   plugins/UFDiMuonsAnalyzer.h
modified:   python/Analyzer_MC_cff.py
modified:   python/Analyzer_cff.py
```

Tesing now on MC. Segmentation violation after "Inside UFDIMuonsAnalyzer constructor".

**Comment by bortigno** [comment]
The segmentation violation was due to the different way of handling the electron ID from the config (from inputTag to string). Updating this to string, fixed the crash.

Now the program crashes because the json has a different eta binning and naming ( from abseta_pt to pt_eta for example). Fixing that in src/MuonHelper.cc

**Comment by bortigno** [comment]
The main problem to solve now is how to access an array in the json file.
I wrote to Sergio to see if he has a code snippet.

**Comment by bortigno** [comment]
For the moment I leave it like this 

```
bortigno@lxplus624:~/workspace/x2mm16_9413/src/Ntupliser/DiMuons$ git status
On branch issue-75-muon-iso-id-trig-sf-2016-rereco
Your branch is up-to-date with 'origin/issue-75-muon-iso-id-trig-sf-2016-rereco'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   plugins/UFDiMuonsAnalyzer.cc
	modified:   plugins/UFDiMuonsAnalyzer.h
	modified:   python/Analyzer_MC_cff.py
	modified:   python/Analyzer_cff.py
	modified:   src/MuonHelper.cc
	modified:   test/test_ntupliser_mc.py
```

**Comment by bortigno** [comment]
After interacting with Sergio they update the MuonReferenceEfficiencies repository adding new json files without array. What I need are

```
EfficienciesStudies/2016_legacy_rereco/jsonfiles/RunBCDEF_sf_SYS_NUM_MediumID_DEN_genTracks.json
EfficienciesStudies/2016_legacy_rereco/jsonfiles/Run*_sf_SYS_NUM_LooseRelIso_DEN_MediumID.json
```

Need to update the cff and test again.

**Comment by bortigno** [comment]
Now it is working fine. A small detail to polish would be to check wether the name of the file and the name of the working points coincide, but I don't think it is worth it. In any case the code would crash or throw an exception. Last work on 2c90f12.

**Comment by bortigno** [comment]
Now I want to merge back to `issue-50-l1-prefiring-2016`.
I think it should be possible without problems.

```
git branch
  issue-50-l1-prefiring-2016
  issue-73-backport-2017-18-to-2016
* issue-75-muon-iso-id-trig-sf-2016-rereco
  master_2016_94X
  master_2017_94X

git diff --name-only issue-50-l1-prefiring-2016
DiMuons/plugins/UFDiMuonsAnalyzer.cc
DiMuons/plugins/UFDiMuonsAnalyzer.h
DiMuons/python/Analyzer_MC_cff.py
DiMuons/python/Analyzer_cff.py
DiMuons/src/MuonHelper.cc
```

My last question is wether is would be better to handle it from a PR... I am gonna have lunch and then come back to it.

**Comment by bortigno** [comment]
Merged in #76
